verification_token/models.py

```diff
@@ -19,14 +19,27 @@ class VerificationTokenManager(models.Manager):
     def deactivate(self, obj, slug=None, key=None):
         self.filter_active_tokens(obj, slug, key).update(is_active=False)
 
-    def deactivate_and_create(self, obj, slug=None, extra_data=None, deactivate_old_tokens=True, expiration_in_minutes=None,
-                              key_generator_kwargs=None):
-        expiration_in_minutes = settings.DEFAULT_EXPIRATION if expiration_in_minutes is None else expiration_in_minutes
-
-        key_generator_kwargs = {} if key_generator_kwargs is None else key_generator_kwargs
+    def deactivate_and_create(self, obj, slug=None, extra_data=None, deactivate_old_tokens=True,
+                              key_generator_kwargs=None, **kwargs):
         if deactivate_old_tokens:
             self.deactivate(obj, slug)
 
+        return self._create(obj, slug=slug, extra_data=extra_data, key_generator_kwargs=key_generator_kwargs, **kwargs)
+
+    def get_active_or_create(self, obj, slug=None, extra_data=None, key=None, key_generator_kwargs=None, **kwargs):
+        token = self.filter_active_tokens(obj, slug, key).order_by('created_at').last()
```

Comment:
Tohle ti vrátí i expirované tokeny, to asi nechceš. Klidně bych upravil tu funkci následovně:
- `filter_active_tokens` -> `filter_valid_tokens`
- kontrolovala by i jestli je token expirovaný

Druhá možnost je přidat do if níže ještě `and token.is_valid()`, ale obecně mi přijde ten zmiňovaný filtr takový divný, když filtruje pouze aktivní a vrací i expirované, moc v tom nevidím smysl

Reply:
Souhlas, je takový divný. Bohužel je token a jeho platnost konstruována podle mě nešťastně v tom směru, že se špatně (zatím mě nenapadlo jak, mimo raw sql) kontroluje uplynutí platnosti pouze v jednom dotazu. Dělat cyklus by bylo potenciálně vražedné (při použití filtru na větší množství záznamů).

Příklad: platnost se počítá z datetime [created_at] a integeru [expires_in_minutes], takže je potřeba spočítat datetime uplynutí platnosti a to porovnat s aktuálním datetime (now()).

To lze udělat bez problému v Pythonu (create_at + timedelta(minutes=expires_in_minutes) > now()), pak by se ale člověk nevyhnul tomu cyklu kontroly všech záznamů vrácených qs (což by při velkém počtu záznamů mohlo být vražedné).

V raw postgresql sql by to mělo jít udělat zhruba jako `created_at + ((expires_in_minutes + ' minutes')::text)::interval > now()` (a na základě takto vygenerované hodnoty vybírat), ale jak takového konstruktu (zejména těch sql castů) dosáhnout přes ORM jsem zatím nepřišel.

Nebo mi něco uniká? Nějaké nápady?

Reply:
Takové ty `F()` věci z Django ORM nelze použít?

Reply:
hmm vidíš toho jsem si nevšiml. Raw nedělat ani za zlaté prase. Ten filtr je fakt divný zasloužil by si opravit. Jak říká ondra s F to uděláš dost snadno

Reply:
Zas tak snadné to není (nebo,jak jsem psal výše, mi něco uniká), jediné na co jsem zatím přišel a co projde je:
`return self.filter(created_at__gte=Now() - Cast(Concat(Cast('expiration_in_minutes', TextField()), Value(' minutes'), output_field=TextField()), DurationField()))`, ale zatím to nemám otestované.

Reply:
Jinak F() moc přímo použít nejde, protože timedelta potřebuje jako vstup integer hodnotu, ale ta je v db poli (a F objekt nezkousne) - leda zkusit cast jako výše.

Reply:
S castem to samozřejmě taky nejde... Takže mi z toho nakonec vylezlo:
`return self.filter(
            Q(expiration_in_minutes=None) |
            Q(created_at__gte=Now() - Cast(
                Concat(
                    Cast(
                        Coalesce('expiration_in_minutes', Value(0))
                        , TextField()),
                    Value(' minutes'),
                    output_field=TextField()
                ),
                DurationField())
            )
        )
`

Reply:
stačí trošku googlit ThatModel.objects.filter(
    last_datetime__lte=now + datetime.timedelta(seconds=1)*F("interval"))

Reply:
tedy pro tebe

self.filger(Q(expiration_in_minutes=None) | Q(now - datetime.timedelta(minutes=1) * F('expiration_in_minutes))

Reply:
Po telefonické dohodě s @matllubos se použije druhé řešení s kontroloou tokenu na .is_valid, protože řešení s timedelta není přenosné na všechny databáze a mnou navrhované řešení (cast, concat atd.) trpí stejným problémem.
Udělá se technical improvements task na refaktorizaci knihovny v tom smyslu, že se přidá datetime pole konce platnosti tokenu, pak bude srovnání datumů fungovat všude.